velox/exec/HashTable.cpp

```diff
@@ -613,22 +636,36 @@ void HashTable<ignoreNullKeys>::clear() {
 
 template <bool ignoreNullKeys>
 void HashTable<ignoreNullKeys>::checkSize(int32_t numNew) {
+  VELOX_CHECK(
+      size_ == 0 || size_ > (numDistinct_ + numTombstones_),
+      "size_:{} numDistinct_:{} numTombstoneRows_:{}",
+      size_,
+      numDistinct_,
+      numTombstones_);
+
+  const int64_t numDistincts = numNew + numDistinct_;
   if (!table_ || !size_) {
     // Initial guess of cardinality is double the first input batch or at
     // least 2K entries.
     // numDistinct_ is non-0 when switching from HashMode::kArray to regular
     // hashing.
     auto newSize = std::max(
         (uint64_t)2048, bits::nextPowerOfTwo(numNew * 2 + numDistinct_));
-    if (numNew + numDistinct_ > rehashSize(newSize)) {
+    if (numDistincts > rehashSize(newSize)) {
       newSize *= 2;
     }
     allocateTables(newSize);
-    if (numDistinct_) {
+    if (numDistinct_ > 0) {
       rehash();
     }
-  } else if (numNew + numDistinct_ > rehashSize()) {
-    auto newSize = bits::nextPowerOfTwo(size_ + numNew);
+    // We are not always able to reuse a tombstone slot as a free one for hash
```

Comment:
I guess not but I am not quite sure if I understand your question. Because once a slot is not found in a group and there is a free slot the group and we just use that slot to store the entry which might proceeds the target slot position in the group.

Reply:
I was wondering about the case where we delete the last item before empty slot. In this case, we can mark it empty rather than tombstone. Maybe we already do that.